In [1]:
import csv
import sqlite3
import os
import numpy as np
import pandas as pd



In [ ]:
#Opening patient statusinfor
with open('Patient_Status.csv', newline='') as csvfile:
    data = list(csv.reader(csvfile))

In [2]:
#No of HC patients
healthy_patient=[]
for i in range(len(data)):
    if data[i][1]=='HC':
        healthy_patient.append(int(data[i][0]))
        
len(healthy_patient)

241

In [3]:
#Opening blood_chem_data
with open('Blood_Chemistry___Hematology.csv', newline='') as csvfile:
    blood_data = list(csv.reader(csvfile))

[['PATNO',
  'EVENT_ID',
  'PAG_NAME',
  'LCOLLDT',
  'COLLTM',
  'LRECDT',
  'RECTM',
  'LRPTDT',
  'RPTTM',
  'LABCODE',
  'LGROUP',
  'LTSTCODE',
  'LTSTNAME',
  'LVISTYPE',
  'LSIRES',
  'LSIUNIT',
  'LSILORNG',
  'LSIHIRNG',
  'LUSRES',
  'LUSUNIT',
  'LUSLORNG',
  'LUSHIRNG',
  'LRESFLG',
  'LTSTCOMM'],
 ['3404',
  'SC',
  'COVANCE',
  '06/2010',
  '10:30:00',
  '06/2010',
  '04:56:00',
  '06/2010',
  '22:14:00',
  'COVANCE',
  'COAGU',
  'CGT283',
  'Prothrombin Time',
  'Screening',
  '10.1',
  'sec',
  '9.7',
  '12.3',
  '10.1',
  'sec',
  '9.7',
  '12.3',
  '',
  ''],
 ['3404',
  'SC',
  'COVANCE',
  '06/2010',
  '10:30:00',
  '06/2010',
  '04:56:00',
  '06/2010',
  '22:14:00',
  'COVANCE',
  'COAGU',
  'CGT284',
  'APTT-QT',
  'Screening',
  '21.0',
  'sec',
  '22.8',
  '31.0',
  '21.0',
  'sec',
  '22.8',
  '31.0',
  'L',
  ''],
 ['3404',
  'SC',
  'COVANCE',
  '06/2010',
  '10:30:00',
  '06/2010',
  '04:56:00',
  '06/2010',
  '17:28:00',
  'COVANCE',
  'HEMAT',
  'HMT10',


In [20]:
#List of blood tests
test=[]
for i in range(1,34):
    test.append(blood_data[i][12])
test

['Prothrombin Time',
 'APTT-QT',
 'Monocytes',
 'Eosinophils',
 'Basophils',
 'Platelets',
 'Neutrophils (%)',
 'Lymphocytes (%)',
 'Monocytes (%)',
 'Eosinophils (%)',
 'Basophils (%)',
 'Hematocrit',
 'RBC',
 'Hemoglobin',
 'WBC',
 'RBC Morphology',
 'Neutrophils',
 'Lymphocytes',
 'Total Bilirubin',
 'Serum Glucose',
 'Total Protein',
 'Albumin-QT',
 'Alkaline Phosphatase-QT',
 'Serum Sodium',
 'Serum Potassium',
 'Serum Bicarbonate',
 'Serum Chloride',
 'Calcium (EDTA)',
 'Creatinine (Rate Blanked)',
 'ALT (SGPT)',
 'AST (SGOT)',
 'Urea Nitrogen',
 'Serum Uric Acid']

In [4]:
#Patient IDs for healthy and unhealty groups
healthy_blood_data=[]
unique_healthy_db=[]
unique_unhealthy_db=[]
for i in range(1,len(blood_data)):
    if int(blood_data[i][0]) in healthy_patient:
        healthy_blood_data.append(blood_data[i])
        if int(blood_data[i][0]) not in unique_healthy_db:
            unique_healthy_db.append(int(blood_data[i][0]))
    else:
        if int(blood_data[i][0]) not in unique_unhealthy_db:
            unique_unhealthy_db.append(int(blood_data[i][0]))

In [5]:
healthy_blood_data
len(unique_healthy_db)

232

In [16]:
unique_bd_patient_ids=unique_unhealthy_db+unique_healthy_db

Putin study:

- 241 HC

- 232 have had the blood tests

- These include 4/5 results that the Putin study found to be most significant for age prediction 

- We considered as an initial step to work on the blood works data etc, potentially following the work of the Putin study and then integrate MRI images into this

In [6]:
data = pd.read_csv ('Blood_Chemistry___Hematology.csv')   
blood_chem_df = pd.DataFrame(data)

def filter_rows_by_values(df, col, values):
    return df[df[col].isin(values) == True]

#blood_chem_df=filter_rows_by_values(blood_chem_df,"PATNO", healthy_patient)
blood_chem_df=blood_chem_df.set_index("PATNO")
blood_chem_df


EVENT_ID PAG_NAME  LCOLLDT    COLLTM   LRECDT     RECTM   LRPTDT  \
PATNO                                                                    
3404        SC  COVANCE  06/2010  10:30:00  06/2010  04:56:00  06/2010   
3404        SC  COVANCE  06/2010  10:30:00  06/2010  04:56:00  06/2010   
3404        SC  COVANCE  06/2010  10:30:00  06/2010  04:56:00  06/2010   
3404        SC  COVANCE  06/2010  10:30:00  06/2010  04:56:00  06/2010   
3404        SC  COVANCE  06/2010  10:30:00  06/2010  04:56:00  06/2010   
...        ...      ...      ...       ...      ...       ...      ...   
60134       SC  COVANCE  10/2014  10:00:00  10/2014  07:00:00      NaN   
60134       SC  COVANCE  10/2014  10:00:00  10/2014  07:00:00      NaN   
60134       SC  COVANCE  10/2014  10:00:00  10/2014  07:00:00      NaN   
60134       SC  COVANCE  10/2014  10:00:00  10/2014  07:00:00      NaN   
60134       SC  COVANCE  10/2014  10:00:00  10/2014  07:00:00      NaN   

          RPTTM  LABCODE LGROUP  ... LSIRES LSIUNIT LSILORNG LSIHIRNG LUSRES  \
PATNO                            ...                                           
3404   22:14:00  COVANCE  COAGU  ...   10.1     sec     9.70    12.30   10.1   
3404   22:14:00  COVANCE  COAGU  ...   21.0     sec    22.80    31.00   21.0   
3404   17:28:00  COVANCE  HEMAT  ...   0.28    GI/L     0.12     0.92   0.28   
3404   17:28:00  COVANCE  HEMAT  ...   0.09    GI/L     0.00     0.57   0.09   
3404   17:28:00  COVANCE  HEMAT  ...   0.03    GI/L     0.00     0.20   0.03   
...         ...      ...    ...  ...    ...     ...      ...      ...    ...   
60134       NaN  COVANCE  CHEMI  ...    NaN  umol/L      NaN      NaN    NaN   
60134       NaN  COVANCE  CHEMI  ...    NaN  umol/L      NaN      NaN    NaN   
60134       NaN  COVANCE  CHEMI  ...    NaN     g/L      NaN      NaN    NaN   
60134       NaN  COVANCE  CHEMI  ...    NaN  mmol/L      NaN      NaN    NaN   
60134       NaN  COVANCE  HEMAT  ...    NaN    GI/L      NaN      NaN    NaN   

        LUSUNIT  LUSLORNG LUSHIRNG LRESFLG                    LTSTCOMM  
PATNO                                                                   
3404        sec      9.70    12.30     NaN                         NaN  
3404        sec     22.80    31.00       L                         NaN  
3404   x10^3/uL      0.12     0.92     NaN                         NaN  
3404   x10^3/uL      0.00     0.57     NaN                         NaN  
3404   x10^3/uL      0.00     0.20     NaN                         NaN  
...         ...       ...      ...     ...                         ...  
60134     mg/dL       NaN      NaN     NaN  discarded per instructions  
60134     mg/dL       NaN      NaN     NaN  discarded per instructions  
60134      g/dL       NaN      NaN     NaN  discarded per instructions  
60134     mg/dL       NaN      NaN     NaN  discarded per instructions  
60134  x10^3/uL       NaN      NaN     NaN  discarded per instructions  

[217512 rows x 23 columns]

In [7]:
screen_dem_data = pd.read_csv ('Screening___Demographics.csv')   
screen_dem_df = pd.DataFrame(screen_dem_data)
    
screen_dem_df=screen_dem_df.set_index('PATNO')


1971.0

## Extracting data for EDA

In [24]:
clean_blood_df = pd.DataFrame({"PATNO":unique_bd_patient_ids})
clean_blood_df=clean_blood_df.set_index('PATNO')
total=len(unique_bd_patient_ids)
clean_blood_df['YOB']=np.zeros(total)
clean_blood_df['AGE']=np.zeros(total)
for i in unique_bd_patient_ids:
    try:
        clean_blood_df.loc[i]['YOB']=screen_dem_df.loc[i]['BIRTHDT']
        clean_blood_df.loc[i]['AGE']=2010-screen_dem_df.loc[i]['BIRTHDT']
    except KeyError:
        clean_blood_df.loc[i]['YOB']="NaN"
        clean_blood_df.loc[i]['AGE']="NaN"
    


In [36]:
clean_blood_df['HEALTHY']=np.zeros(total)

for i in unique_bd_patient_ids:
    if i in unique_healthy_db:
        clean_blood_df.loc[i]['HEALTHY']=1
    else:
        clean_blood_df.loc[i]['HEALTHY']=0


In [37]:
clean_blood_df


YOB   AGE  HEALTHY
PATNO                       
3400   1971.0  39.0      0.0
3403   1941.0  69.0      0.0
3402   1964.0  46.0      0.0
3406   1975.0  35.0      0.0
3407   1945.0  65.0      0.0
...       ...   ...      ...
3171   1950.0  60.0      1.0
3157   1946.0  64.0      1.0
3191   1947.0  63.0      1.0
3172   1942.0  68.0      1.0
4105   1946.0  64.0      1.0

[2014 rows x 3 columns]

In [38]:
for j in test:
    clean_blood_df[j]=np.zeros(total)

In [39]:
clean_blood_df

YOB   AGE  HEALTHY  Prothrombin Time  APTT-QT  Monocytes  \
PATNO                                                                
3400   1971.0  39.0      0.0               0.0      0.0        0.0   
3403   1941.0  69.0      0.0               0.0      0.0        0.0   
3402   1964.0  46.0      0.0               0.0      0.0        0.0   
3406   1975.0  35.0      0.0               0.0      0.0        0.0   
3407   1945.0  65.0      0.0               0.0      0.0        0.0   
...       ...   ...      ...               ...      ...        ...   
3171   1950.0  60.0      1.0               0.0      0.0        0.0   
3157   1946.0  64.0      1.0               0.0      0.0        0.0   
3191   1947.0  63.0      1.0               0.0      0.0        0.0   
3172   1942.0  68.0      1.0               0.0      0.0        0.0   
4105   1946.0  64.0      1.0               0.0      0.0        0.0   

       Eosinophils  Basophils  Platelets  Neutrophils (%)  ...  Serum Sodium  \
PATNO                                                      ...                 
3400           0.0        0.0        0.0              0.0  ...           0.0   
3403           0.0        0.0        0.0              0.0  ...           0.0   
3402           0.0        0.0        0.0              0.0  ...           0.0   
3406           0.0        0.0        0.0              0.0  ...           0.0   
3407           0.0        0.0        0.0              0.0  ...           0.0   
...            ...        ...        ...              ...  ...           ...   
3171           0.0        0.0        0.0              0.0  ...           0.0   
3157           0.0        0.0        0.0              0.0  ...           0.0   
3191           0.0        0.0        0.0              0.0  ...           0.0   
3172           0.0        0.0        0.0              0.0  ...           0.0   
4105           0.0        0.0        0.0              0.0  ...           0.0   

       Serum Potassium  Serum Bicarbonate  Serum Chloride  Calcium (EDTA)  \
PATNO                                                                       
3400               0.0                0.0             0.0             0.0   
3403               0.0                0.0             0.0             0.0   
3402               0.0                0.0             0.0             0.0   
3406               0.0                0.0             0.0             0.0   
3407               0.0                0.0             0.0             0.0   
...                ...                ...             ...             ...   
3171               0.0                0.0             0.0             0.0   
3157               0.0                0.0             0.0             0.0   
3191               0.0                0.0             0.0             0.0   
3172               0.0                0.0             0.0             0.0   
4105               0.0                0.0             0.0             0.0   

       Creatinine (Rate Blanked)  ALT (SGPT)  AST (SGOT)  Urea Nitrogen  \
PATNO                                                                     
3400                         0.0         0.0         0.0            0.0   
3403                         0.0         0.0         0.0            0.0   
3402                         0.0         0.0         0.0            0.0   
3406                         0.0         0.0         0.0            0.0   
3407                         0.0         0.0         0.0            0.0   
...                          ...         ...         ...            ...   
3171                         0.0         0.0         0.0            0.0   
3157                         0.0         0.0         0.0            0.0   
3191                         0.0         0.0         0.0            0.0   
3172                         0.0         0.0         0.0            0.0   
4105                         0.0         0.0         0.0            0.0   

       Serum Uric Acid  
PATNO                   
3400               0.0  
3403 

In [40]:
for i in unique_bd_patient_ids:

    tmp=pd.DataFrame(blood_chem_df.loc[i])
    tmp=tmp.set_index('LTSTNAME')
    

    for j in test:
        try:
            v=tmp.loc[j]['LSIRES']
            try:
                v.dtype
                clean_blood_df.loc[i][j]=v[-1]
            except AttributeError:
                clean_blood_df.loc[i][j]=v
        except KeyError:
            clean_blood_df.loc[i][j]="N/A"
        

In [41]:
clean_blood_df

YOB   AGE  HEALTHY  Prothrombin Time  APTT-QT  Monocytes  \
PATNO                                                                
3400   1971.0  39.0      0.0               9.6     25.6       0.41   
3403   1941.0  69.0      0.0              10.6     22.4       0.55   
3402   1964.0  46.0      0.0              10.5     26.0       0.44   
3406   1975.0  35.0      0.0              11.6     20.8       0.20   
3407   1945.0  65.0      0.0              10.3     24.1       0.49   
...       ...   ...      ...               ...      ...        ...   
3171   1950.0  60.0      1.0              11.2     22.5       0.25   
3157   1946.0  64.0      1.0              10.0     22.8       0.50   
3191   1947.0  63.0      1.0              10.6     21.8       0.20   
3172   1942.0  68.0      1.0              10.8     25.1       0.41   
4105   1946.0  64.0      1.0              10.2     27.4       0.38   

       Eosinophils  Basophils  Platelets  Neutrophils (%)  ...  Serum Sodium  \
PATNO                                                      ...                 
3400          0.09       0.05      336.0             79.1  ...         139.0   
3403          0.16       0.20      157.0             10.2  ...         146.0   
3402          0.16       0.04      326.0             63.9  ...         141.0   
3406          0.03       0.01      207.0             65.8  ...         139.0   
3407          0.33       0.02      219.0             67.5  ...         140.0   
...            ...        ...        ...              ...  ...           ...   
3171          0.08       0.01      214.0             49.1  ...         135.0   
3157          0.13       0.02      388.0             64.7  ...         139.0   
3191          0.06       0.05      229.0             53.8  ...         143.0   
3172          0.07       0.03      234.0             67.2  ...         133.0   
4105          0.14       0.06      170.0             59.9  ...         138.0   

       Serum Potassium  Serum Bicarbonate  Serum Chloride  Calcium (EDTA)  \
PATNO                                                                       
3400               4.2               20.2           104.0            2.30   
3403               5.4               23.3           105.0            2.47   
3402               4.1               25.9           100.0            2.50   
3406               4.2               26.2            99.0            2.35   
3407               3.7               22.6           101.0            2.59   
...                ...                ...             ...             ...   
3171               4.2               23.0            98.0            2.42   
3157               4.3               24.3           101.0            2.42   
3191               3.9               26.1           101.0            2.37   
3172               3.8               20.7           100.0            2.27   
4105               4.5               24.3           100.0            2.30   

       Creatinine (Rate Blanked)  ALT (SGPT)  AST (SGOT)  Urea Nitrogen  \
PATNO                                                                     
3400                        80.0        19.0        15.0            7.1   
3403                        71.0        25.0        23.0            6.8   
3402                        80.0        26.0        21.0            5.4   
3406                        80.0        17.0        19.0            3.9   
3407                        88.0        26.0        44.0            6.8   
...                          ...         ...         ...            ...   
3171                        97.0        24.0        29.0            6.4   
3157                        71.0        22.0        25.0            8.2   
3191                        88.0        23.0        26.0            4.3   
3172                        71.0        22.0        21.0            4.6   
4105                        88.0        26.0        26.0            4.6   

       Serum Uric Acid  
PATNO                   
3400             268.0  
3403 

In [42]:
clean_blood_df.to_csv('clean_blood_test_all.csv', encoding='utf-8', index=True)